get_ipython().run_cell_magic('bash', '', '\nBASE_PATH=/hackathon/Hackathon_Project_4\ncd $BASE_PATH/GENOME_ANNOTATIONS/\n\nawk -F "\\t" \'{if ($3 == "CDS") print $0}\' gencode.v19.annotation.chr22.gtf | \\\nawk \'{print $1 "\\t" $4 "\\t" $5 "\\t" $28"-"$18 "\\t" "." "\\t" $7}\' | \\\nawk \'{gsub(/;/,"");print}\' | awk \'{gsub(/"/,"");print}\' > CDSs.chr22.bed\n\nawk -F "\\t" \'{if ($3 == "UTR") print $0}\' gencode.v19.annotation.chr22.gtf | \\\nawk \'{print $1 "\\t" $4 "\\t" $5 "\\t" $12"-"$18 "\\t" "." "\\t" $7}\' | \\\nawk \'{gsub(/;/,"");print}\' | awk \'{gsub(/"/,"");print}\' > UTRs.chr22.bed\n\nawk -F "\\t" \'{if ($3 == "exon") print $0}\' gencode.v19.annotation.chr22.gtf | \\\nawk \'{print $1 "\\t" $4 "\\t" $5 "\\t" $28"-"$18 "\\t" "." "\\t" $7}\' | \\\nawk \'{gsub(/;/,"");print}\' | awk \'{gsub(/"/,"");print}\' > Exons.chr22.bed\n\nawk -F "\\t" \'{if ($3 == "gene") print $0}\' gencode.v19.annotation.chr22.gtf | \\\nawk \'{print $1 "\\t" $4 "\\t" $5 "\\t" $10"-"$18 "\\t" "." "\\t" $7}\' | \\\nawk \'{gsub(/;/,"");print}\' | awk \'{gsub(/"/,"");print}\' > Genes.chr22.bed\n\nawk \'{if ($3 == "transcript") print $1 "\\t" $4 "\\t" $5 "\\t" $12 "\\t" "." "\\t" $7 \\\n"\\t" $10 "\\t" $18 "\\t" $14}\' gencode.v19.annotation.chr22.gtf | awk \'{gsub(/;/,"");print}\' | \\\nawk \'{gsub(/"/,"");print}\' > Transcripts.chr22.bed\n\nawk \'{if ($6 == "+") print $1 "\\t" $2-2000 "\\t" $2+2000 "\\t" $4 "\\t" $5 "\\t" $6 \\\n"\\t" $7 "\\t" $8 "\\t" $9 ; else print $1 "\\t" $3-2000 "\\t" $3+2000 "\\t" $4 "\\t" $5 "\\t" \\\n$6 "\\t" $7 "\\t" $8 "\\t" $9}\' Transcripts.chr22.bed > Promoters.chr22.bed ')

get_ipython().run_cell_magic('bash', '', '\nBASE_PATH=/hackathon/Hackathon_Project_4\ncd $BASE_PATH/GIAB_DATA\n\n$BASE_PATH/SOFTWARE/Benchmark-Tools/Determine-Variant-Context.sh \\\nGIAB-V3.3-chr22.bed GIAB-V3.3-chr22.Annotated.bed')

get_ipython().run_cell_magic('bash', '', 'source ~/.bashrc\nBASE_PATH=/hackathon/Hackathon_Project_4\ncd $BASE_PATH/GIAB_DATA\n\nbedtools intersect -u -b GIAB-V3.3-chr22.bed -a $BASE_PATH/ENCODE_DATA_GM12878/PEAKS/ENCFF002CDP.bed | \\\nsort -k7,7rg | head -50 > peaks.bed\n\necho -e "chrom\\tposition\\tposition\\tsnp_id\\tgenotype\\tmajor_allele\\tminor_allele\\tdepth\\tgenotype_quality" \\\n> $BASE_PATH/BENCHMARK/Benchmark-Test-V1.txt\n\nbedtools intersect -u -a GIAB-V3.3-chr22.bed -b peaks.bed | \\\nsort -k9,9rg | head -25 >> $BASE_PATH/BENCHMARK/Benchmark-Test-V1.txt\n\nrm peaks.bed\nhead $BASE_PATH/BENCHMARK/Benchmark-Test-V1.txt | column -t')



